### PR TITLE
Implement `runIf()` function

### PR DESCRIPTION
**Pull request by o-rumiantsev**
Closes: https://github.com/metarhia/metasync/issues/410

### CODE REVIEW

File: test/control.js [hunk 1]
```diff
@@ -79,3 +79,48 @@ metatests.test('sequential with error', test => {
...
+metatests.test('runIf run asyncFn', test => {
+  const condition = true;
+  const asyncFn = (arg1, arg2, cb) => {
+    process.nextTick(() => {
+      cb(null, { arg1, arg2 });
+    });
+  };
+
+  metasync.runIf(condition, asyncFn, 'val1', 'val2', (err, result) => {
+    test.strictSame(err, null);
```
Comment (SemenchenkoVitaliy) [inline test/control.js]:
```suggestion
    test.error(err);
```

File: test/control.js [hunk 2]
```diff
@@ -79,3 +79,48 @@ metatests.test('sequential with error', test => {
...
+metatests.test('runIf do not run asyncFn', test => {
+  const condition = false;
+  const asyncFn = (arg1, arg2, cb) => {
+    process.nextTick(() => {
+      cb(null, { arg1, arg2 });
+    });
+  };
+
+  metasync.runIf(condition, asyncFn, 'val1', 'val2', (err, result) => {
+    test.strictSame(err, null);
```
Comment (SemenchenkoVitaliy) [inline test/control.js]:
```suggestion
    test.error(err);
```

File: test/control.js [hunk 3]
```diff
@@ -79,3 +79,48 @@ metatests.test('sequential with error', test => {
...
+metatests.test('runIf default value', test => {
+  const condition = false;
+  const asyncFn = (val, cb) => {
+    process.nextTick(() => {
+      cb(null, val);
+    });
+  };
+
+  metasync.runIf(condition, 'default', asyncFn, 'val', (err, result) => {
+    test.strictSame(err, null);
```
Comment (SemenchenkoVitaliy) [inline test/control.js]:
```suggestion
    test.error(err);
```

File: test/control.js [hunk 4]
```diff
@@ -79,3 +79,48 @@ metatests.test('sequential with error', test => {
...
+
+metatests.test('runIf run asyncFn', test => {
+  const condition = true;
+  const asyncFn = (arg1, arg2, cb) => {
+    process.nextTick(() => {
+      cb(null, { arg1, arg2 });
+    });
+  };
+
+  metasync.runIf(condition, asyncFn, 'val1', 'val2', (err, result) => {
```
Comment (lundibundi) [inline test/control.js]: Could you wrap the callbacks in `test.mustCall` and `test.mustNotCall` where needed?

File: test/control.js [hunk 5]
```diff
@@ -79,3 +79,48 @@ metatests.test('sequential with error', test => {
...
+      cb(null, val);
+    });
+  };
+
+  metasync.runIf(condition, 'default', asyncFn, 'val', (err, result) => {
+    test.error(err);
+    test.strictSame(result, 'default');
+    test.end();
+  });
+});
```
Comment (lundibundi) [inline test/control.js]: Add a test that `runIf` forwards an error from `asyncFn`.

File: test/control.js [hunk 6]
```diff
@@ -79,3 +79,48 @@ metatests.test('sequential with error', test => {
...
+
+metatests.test('runIf default value', test => {
+  const condition = false;
+  const asyncFn = (val, cb) => {
+    process.nextTick(() => {
+      cb(null, val);
+    });
+  };
+
+  metasync.runIf(condition, 'default', asyncFn, 'val', (err, result) => {
```
Comment (lundibundi) [inline test/control.js]: Nit: IMO it's clearer if `condition` is inlined in these test, i.e. `metasync.runIf(false, ...)`.

File: test/control.js [hunk 7]
```diff
@@ -79,3 +79,58 @@ metatests.test('sequential with error', test => {
...
+
+  metasync.runIf(false, asyncFn, 'val1', 'val2', (err, result) => {
+    test.error(err);
+    test.strictSame(result, undefined);
+    test.end();
+  });
+});
+
+metatests.test('runIf default value', test => {
+  const asyncFn = (val, cb) => {
```
Comment (lundibundi) [inline test/control.js]: I think this can just be `const asyncFn = test.mustNotCall()`.

File: test/control.js [hunk 8]
```diff
@@ -79,3 +79,58 @@ metatests.test('sequential with error', test => {
...
+
+  metasync.runIf(true, asyncFn, 'val1', 'val2', (err, result) => {
+    test.error(err);
+    test.strictSame(result, { arg1: 'val1', arg2: 'val2' });
+    test.end();
+  });
+});
+
+metatests.test('runIf do not run asyncFn', test => {
+  const asyncFn = test.mustNotCall((arg1, arg2, cb) => {
```
Comment (lundibundi) [inline test/control.js]: And same here. `const asyncFn = test.mustNotCall()`, wdyt?

Comment (o-rumiantsev) [reply test/control.js]: This will not work, because `asyncFn` must call callback inside.

Comment (lundibundi) [reply test/control.js]: You mean it doesn't have a proper signature of an async function? Well, it looks ok to me though if you think it'll be clearer with verbose version feel free to leave it as is.

File: test/control.js (hunk 7 above)
Comment (lundibundi) [reply test/control.js]: ping @o-rumiantsev

Comment (o-rumiantsev) [reply test/control.js]: Sorry. I think it is better to leave it as it is now.

Comment (nechaido) [reply test/control.js]: @o-rumiantsev why?

Comment (o-rumiantsev) [reply test/control.js]: @nechaido think, this can be less understandable

Comment (belochub) [reply test/control.js]: @o-rumiantsev can you at least wrap this function in `test.mustNotCall()` like you did in the test above this one (`'runIf do not run asyncFn'`)?

Comment (o-rumiantsev) [reply test/control.js]: @belochub ofc

File: lib/control.js [hunk 9]
```diff
@@ -98,8 +98,23 @@ const sequential = (fns, context, callback) => {
...
+  if (typeof defaultVal === 'function') {
+    args.unshift(asyncFn);
+    asyncFn = defaultVal;
+    defaultVal = undefined;
+  }
+  if (condition) {
+    asyncFn(...args);
+  } else {
+    const callback = common.last(args);
+    callback(null, defaultVal);
```
Comment (SemenchenkoVitaliy) [inline lib/control.js]: Why not `common.safeCallback`?

File: lib/control.js [hunk 10]
```diff
@@ -98,8 +98,23 @@ const sequential = (fns, context, callback) => {
   next();
 };
 
+const runIf = (condition, defaultVal, asyncFn, ...args) => {
```
Comment (belochub) [inline lib/control.js]: Just a nit: I think we should also add a documentation comment above the function.

File: lib/control.js (hunk 9 above)
Comment (o-rumiantsev) [reply lib/control.js]: Forgot about this feature. Rewritten now.

File: lib/control.js [hunk 11]
```diff
@@ -98,8 +98,30 @@ const sequential = (fns, context, callback) => {
...
+  if (typeof defaultVal === 'function') {
+    args.unshift(asyncFn);
+    asyncFn = defaultVal;
+    defaultVal = undefined;
+  }
+  if (condition) {
+    asyncFn(...args);
+  } else {
+    const callback = common.safeCallback(args);
```
Comment (belochub) [inline lib/control.js]: What would be the point of calling this function without a callback?

File: lib/control.js [hunk 12]
```diff
@@ -98,8 +98,30 @@ const sequential = (fns, context, callback) => {
   next();
 };
 
+// Run `asyncFn` if `condition` is truthy, else return `defaultVal` to callback.
+//   condition - <any>,
```
Comment (belochub) [inline lib/control.js]:
```suggestion
//   condition - <any>
```

File: lib/control.js (hunk 11 above)
Comment (o-rumiantsev) [reply lib/control.js]: Maybe, synchronous function passed as `asyncFn`.

Comment (belochub) [reply lib/control.js]: @o-rumiantsev, in that case calling `metasync.runIf(condition, asyncFn, ...args)` can be replaced with
```js
if (condition) {
  asyncFn(...args); // synchronous function
}
```

Comment (o-rumiantsev) [reply lib/control.js]: So, should I revert this change?

Comment (nechaido) [reply lib/control.js]: You probably should.

File: lib/control.js [hunk 13]
```diff
@@ -98,8 +98,30 @@ const sequential = (fns, context, callback) => {
   next();
 };
 
+// Run `asyncFn` if `condition` is truthy, else return `defaultVal` to callback.
```
Comment (SemenchenkoVitaliy) [inline lib/control.js]:
```suggestion
// Run `asyncFn` if `condition` is truthy, else return `defaultVal` to callback.
// Signature: condition, defaultVal, asyncFn, ...args
```

File: lib/control.js [hunk 14]
```diff
@@ -98,8 +98,30 @@ const sequential = (fns, context, callback) => {
...
+//   condition - <any>
```
[the remark this answers is not in the document]
Comment (o-rumiantsev) [reply lib/control.js]: I thought it can not only be boolean (e.g `''`, `null`, `undefined`, `NaN`, 0)

[the remark this answers is not in the document]
Comment (belochub) [reply lib/control.js]: It's the same as with the regular `if` statement, you can pass any value, not just boolean.

File: test/control.js [hunk 15]
```diff
@@ -79,3 +79,58 @@ metatests.test('sequential with error', test => {
...
+
+  metasync.runIf(false, 'default', asyncFn, 'val', (err, result) => {
+    test.error(err);
+    test.strictSame(result, 'default');
+    test.end();
+  });
+});
+
+metatests.test('runIf forward an error', test => {
+  const asyncFn = cb => {
```
Comment (lundibundi) [inline test/control.js]: `test.mustCall`

**Pull request merged**